core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_services.go

```diff
@@ -123,25 +124,48 @@ func (builtin *AddServicesCapabilities) Execute(ctx context.Context, _ *builtin_
 		renderedServiceConfigs[renderedServiceName] = renderedServiceConfig
 	}
 
-	startedServices, failedServices, err := builtin.serviceNetwork.AddServices(ctx, renderedServiceConfigs, parallelism)
+	serviceToUpdate := map[service.ServiceName]*service.ServiceConfig{}
+	serviceToCreate := map[service.ServiceName]*service.ServiceConfig{}
+	for serviceName, serviceConfig := range renderedServiceConfigs {
+		if _, found := builtin.serviceNetwork.GetServiceRegistration(serviceName); found {
+			serviceToUpdate[serviceName] = serviceConfig
+		} else {
+			serviceToCreate[serviceName] = serviceConfig
+		}
+	}
+
+	updatedServices, failedToBeUpdatedServices, err := builtin.serviceNetwork.UpdateServices(ctx, serviceToUpdate, parallelism)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Unexpected error occurred updating a batch of services")
+	}
+
+	startedServices, failedToBeStartedServices, err := builtin.serviceNetwork.AddServices(ctx, renderedServiceConfigs, parallelism)
 	if err != nil {
 		return "", stacktrace.Propagate(err, "Unexpected error occurred starting a batch of services")
 	}
-	if len(failedServices) > 0 {
-		failedServiceNames := make([]service.ServiceName, len(failedServices))
-		idx := 0
-		for failedServiceName := range failedServices {
-			failedServiceNames[idx] = failedServiceName
-			idx++
+	if len(failedToBeStartedServices) > 0 || len(failedToBeUpdatedServices) > 0 {
+		var failedServiceNames []service.ServiceName
+		for failedServiceName := range failedToBeStartedServices {
+			failedServiceNames = append(failedServiceNames, failedServiceName)
+		}
+		for failedServiceName := range failedToBeUpdatedServices {
+			failedServiceNames = append(failedServiceNames, failedServiceName)
```

Comment:
if all services (including the updated one) are added to the AddServices call the `failedServiceNames` could end up having duplicated names.

Reply:
Yup, you're right but we should pass all the services to AddServices here, only the ones that are being created (i.e. not updated)